src/libraries/System.Reflection.Emit/src/System/Reflection/Emit/AssemblyBuilderImpl.cs

```diff
@@ -0,0 +1,129 @@
+﻿// Licensed to the .NET Foundation under one or more agreements.
+// The .NET Foundation licenses this file to you under the MIT license.
+
+using System.Collections.Generic;
+using System.IO;
+using System.Reflection.Metadata;
+using System.Reflection.Metadata.Ecma335;
+using System.Reflection.PortableExecutable;
+
+namespace System.Reflection.Emit
+{
+    internal sealed class AssemblyBuilderImpl : AssemblyBuilder
+    {
+        private bool _previouslySaved;
+        private readonly AssemblyName _assemblyName;
+        private ModuleBuilderImpl? _module;
+
+        internal AssemblyBuilderImpl(AssemblyName name, IEnumerable<CustomAttributeBuilder>? assemblyAttributes)
+        {
+            ArgumentNullException.ThrowIfNull(name);
+
+            _assemblyName = name;
+
+            if (assemblyAttributes != null)
+            {
+                foreach (CustomAttributeBuilder assemblyAttribute in assemblyAttributes)
+                {
+                    SetCustomAttribute(assemblyAttribute);
+                }
+            }
+        }
+
+        internal static AssemblyBuilderImpl DefineDynamicAssembly(AssemblyName name)
+            => new AssemblyBuilderImpl(name, null);
+
+        internal static AssemblyBuilderImpl DefineDynamicAssembly(
+            AssemblyName name,
+            IEnumerable<CustomAttributeBuilder>? assemblyAttributes)
+                => new AssemblyBuilderImpl(name, assemblyAttributes);
+
+        private static void WritePEImage(Stream peStream, MetadataBuilder metadataBuilder, BlobBuilder ilBuilder) // MethodDefinitionHandle entryPointHandle when we have main method.
+        {
+            // Create executable with the managed metadata from the specified MetadataBuilder.
+            var peHeaderBuilder = new PEHeaderBuilder(
+                imageCharacteristics: Characteristics.Dll // Start off with a simple DLL
+                );
+
+            var peBuilder = new ManagedPEBuilder(
+                peHeaderBuilder,
+                new MetadataRootBuilder(metadataBuilder),
+                ilBuilder);
+
+            // Write executable into the specified stream.
+            var peBlob = new BlobBuilder();
+            peBuilder.Serialize(peBlob);
+            peBlob.WriteContentTo(peStream);
+        }
+
+        internal void Save(Stream stream)
+        {
+            ArgumentNullException.ThrowIfNull(stream);
+
+            if (_module == null)
+            {
+                throw new InvalidOperationException(SR.AModuleRequired);
+            }
+
+            if (_previouslySaved) // Cannot save an assembly multiple times. This is consistent with Save() in .Net Framework.
+            {
+                throw new InvalidOperationException(SR.CannotSaveMultipleTimes);
+            }
+
+            if (string.IsNullOrEmpty(_assemblyName.Name))
+            {
+                throw new InvalidOperationException(SR.AssemblyNameCannotBeNull);
+            }
+
```

Comment:
```suggestion
```
It is better to validate this upfront like how it is done in the runtime builder. (See my other feedback.)